c_src/membrane_audio_mix_plugin/mixer.c

```diff
@@ -1,210 +1,246 @@
+    double quotient = (state->sample_min * 1.0) / min;
+    scale(values, values_length, quotient, state);
+  } else if (max > state->sample_max) {
+    double quotient = (state->sample_max * 1.0) / max;
+    scale(values, values_length, quotient, state);
+  }
+
+  int32_t samples_size = 0;
+  for (uint32_t i = 0; i < state->queue_length;
+       ++i, samples_size += state->sample_size) {
+    caps_audio_raw_value_to_sample(state->queue[i], samples + samples_size,
+                                   state->caps);
+  }
+
+  for (uint32_t i = 0; i < values_length;
+       ++i, samples_size += state->sample_size) {
+    caps_audio_raw_value_to_sample(values[i], samples + samples_size,
+                                   state->caps);
+  }
+
+  unifex_free(state->queue);
+  state->queue = NULL;
+  state->queue_length = 0;
 }
 
-void add_values(int64_t *values, unsigned int values_length, unsigned char *samples, unsigned int *samples_size, bool is_last_wave, UnifexState *state)
-{
-    bool is_wave_positive = state->is_wave_positive;
-    unsigned int start = 0;
-    unsigned int end = split_values(values, values_length, is_wave_positive);
-    *samples_size = (values_length + state->queue_length) * state->sample_size;
-    samples = unifex_alloc(*samples_size);
-    unsigned int samples_start = 0;
-    while (end < values_length)
-    {
-        get_samples(samples + samples_start, values + start, end - start, state);
-        samples_start += (end - start + state->queue_length) * state->sample_size;
-        start = end;
-        state->is_wave_positive = !is_wave_positive;
-        end = split_values(values, values_length, is_wave_positive);
-    }
+/**
+ * Cuts given values on every sign change and convert it to samples.
+ *
+ * Any remaining values are put into state's queue.
+ */
+void cut_values(int64_t *values, uint32_t values_length, uint8_t *samples,
+                uint32_t *samples_length, UnifexState *state) {
+  if (values_length == 0 || !values) {
+    return;
+  }
+
+  bool is_wave_positive = state->is_wave_positive;
+  uint32_t start = 0;
+  uint32_t end = next_sign_change(values, values_length, is_wave_positive);
+
+  uint32_t samples_size = 0;
+  while (end < values_length) {
+    uint32_t queue_length = state->queue_length;
+    get_samples(samples + samples_size, values + start, end - start, state);
+    samples_size += (end - start + queue_length) * state->sample_size;
+    start = end;
+    is_wave_positive = !is_wave_positive;
+    end =
+        next_sign_change(values + end, values_length - end, is_wave_positive) +
+        end;
+  }
+
+  state->is_wave_positive = is_wave_positive;
+
+  uint32_t new_queue_length = state->queue_length + end - start;
+  int64_t *new_queue = unifex_alloc(new_queue_length * sizeof(int64_t));
+  memcpy(new_queue, state->queue, state->queue_length * sizeof(int64_t));
+  memcpy(new_queue + state->queue_length, values + start,
+         (end - start) * sizeof(int64_t));
```

Comment:
Don't use the type with `sizeof` but rather a variable for which you allocate the memory, e.g.
```c
int64_t *new_queue = unifex_alloc(new_queue_length * sizeof(*new_queue));
```